samza-kv/src/test/java/org/apache/samza/storage/kv/TestLargeMessageSafeStore.java

```diff
@@ -0,0 +1,180 @@
+import org.apache.samza.serializers.LongSerde;
+import org.apache.samza.serializers.Serde;
+import org.apache.samza.serializers.StringSerde;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Matchers;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.runners.MockitoJUnitRunner;
+
+
+@RunWith(MockitoJUnitRunner.class)
+public class TestLargeMessageSafeStore {
+
+  @Mock
+  KeyValueStore<byte[], byte[]> store;
+  int maxMessageSize = 1024;
+  String storeName = "testStore";
+
+  @Before
+  public void setup() {
+    Mockito.doNothing().when(store).put(Matchers.any(), Matchers.any());
+    Mockito.doNothing().when(store).putAll(Matchers.any());
+  }
+
+  @Test
+  public void testLargeMessagePutWithDropLargeMessageDisabled() {
+    LargeMessageSafeStore largeMessageSafeKeyValueStore = new LargeMessageSafeStore(store, storeName, false, maxMessageSize);
+
+    byte[] key1 = new byte[16];
+    byte[] smallMessage = new byte[32];
+    largeMessageSafeKeyValueStore.put(key1, smallMessage);
+    Mockito.verify(store).put(Matchers.eq(key1), Matchers.eq(smallMessage));
+
+    byte[] key2 = new byte[16];
+    byte[] largeMessage = new byte[maxMessageSize + 1];
+    try {
+      largeMessageSafeKeyValueStore.put(key2, largeMessage);
+      Assert.fail("The test case should have failed due to a large message being passed to the changelog, but it didn't.");
+    } catch (SamzaException e) {
+      Mockito.verifyZeroInteractions(store);
+    }
+  }
+
+  @Test
+  public void testSmallMessagePutAllSuccessWithDropLargeMessageDisabled() {
+    LargeMessageSafeStore largeMessageSafeKeyValueStore = new LargeMessageSafeStore(store, storeName, false, maxMessageSize);
+    byte[] key = new byte[16];
+    byte[] smallMessage = new byte[32];
+
+    List<Entry<byte[], byte[]>> entries = new ArrayList<>();
+    entries.add(new Entry<>(key, smallMessage));
+
+    largeMessageSafeKeyValueStore.putAll(entries);
+    Mockito.verify(store).putAll(Matchers.eq(entries));
+    Mockito.verify(store, Mockito.never()).put(Matchers.any(byte[].class), Matchers.any(byte[].class));
+  }
+
+  @Test
+  public void testLargeMessagePutAllFailureWithDropLargeMessageDisabled() {
+    LargeMessageSafeStore largeMessageSafeKeyValueStore = new LargeMessageSafeStore(store, storeName, false, maxMessageSize);
+    byte[] key = new byte[16];
+    byte[] largeMessage = new byte[maxMessageSize + 1];
+
+    List<Entry<byte[], byte[]>> entries = new ArrayList<>();
+    entries.add(new Entry<>(key, largeMessage));
+
+    try {
+      largeMessageSafeKeyValueStore.putAll(entries);
+      Assert.fail("The test case should have failed due to a large message being passed to the changelog, but it didn't.");
+    } catch (SamzaException e) {
+      Mockito.verifyZeroInteractions(store);
+    }
+  }
+
+  @Test
+  public void testSmallMessagePutWithSerdeAndDropLargeMessageDisabled() {
```

Comment:
It might not be necessary to have these extra "with serde" tests. The interface for the `LargeMessageSafeStore` is already bytes, and you already tested that case above.

Reply:
Rayman had asked me to add these previously, hence I haven't removed them. Let me know if they absolutely need to be removed.

Reply:
You can leave them in if you believe they are useful.
Just keep in mind in the future that unit tests should test against the API of the class. So if two tests are testing the same flow, then one of them might not be necessary. Every test does have to be maintained, so there is some cost in having more tests.

Reply:
Got it. Leaving them as it is and will keep in mind for the future.